Add WRAPPERS keyword parameter to MAKE-THREAD

scymtym opened the pull request:
This is a draft implementation of a feature I was missing on several occasions and would like to discuss. If there is interest, I can improve and document the code.
- Wrappers are somewhat similar to initial bindings but more general: they can be used to bind variables (with clearer semantics than `initial-bindings`), establish handlers, restarts or catch tags or perform side effects (again with clearer semantics than `initial-bindings`)
- The new variable `*default-bindings*` serves as the default value if the parameter is not supplied

I consider this a more powerful and (imo) easier to understand mechanism than initial bindings. The use cases leading me to this idea were:
- Thread support in fiveam: the `is` macro currently only works in the thread executing the `test`. Using the wrappers mechanism, `test` can arrange for threads created within the scope of the test to forward conditions signaled by `is` (or otherwise) to the current test. For a crude, prototypical implementation, see the [fiveam pull request](https://github.com/sionescu/fiveam/pull/27).
- Consider a [generic progress condition](https://github.com/scymtym/more-conditions#tracking-and-reporting-progress-of-operations) (neither a warning nor an error or a serious condition) that can be freely signaled from any operation to convey its progress without disturbing surrounding control flow. It would be nice to collect all conditions of this kind no matter from which thread they were signaled or who created the thread. This could be achieved by globally pushing a suitable wrapper onto `*default-wrappers*`.

sionescu commented:
More powerful, yes, but not easier to understand, at least not to debug. It's difficult to inspect a closure to see how it affects the spawned threads.
I don't want to make Bordeaux-threads anything more that the building blocks for concurrency. If you want to add more semantics, you can do it in a separate library.

scymtym commented:
> I don't want to make Bordeaux-threads anything more that the building blocks for concurrency. If you want to add more semantics, you can do it in a separate library.

IMO, the proposed wrappers make for a more suitable building block than the current initial bindings: the latter can be implemented in terms of the former, but not the other way around.

That said, I accept your decision and won't pursue this issue further.

sionescu commented:
Does d12e37fb1 work for you ?

scymtym commented:
Thank you for reconsidering, I really appreciate it.

> Does d12e37f work for you ?

I will extend d12e37f sufficiently to attack my use case (e.g. apply the dynamic context in `make-thread`) and report back. Does that sound good to you?

sionescu commented:
Make-thread already applies the dynamic context. See branch "context". I haven't merged it into master so there's time to make changes.

scymtym commented:
> Make-thread already applies the dynamic context. See branch "context"

Yes, my mistake, sorry. I'm currently experimenting with d12e37f and had to change a few things:

* `binding-defaults` called `dynamic-context-wrapper` without an argument.
* `with-dynamic-context` evaluated `(dynamic-context-wrapper *dynamic-context*)` at macroexpansion-time.
* I'm not sure whether nesting `with-dynamic-context` works

I couldn't make my use-case work with the design proposed in d12e37f yet, but I will try a few more things.

> I haven't merged it into master so there's time to make changes.

Thank you. I will report the results of my experiments soon.

scymtym commented:
OK, I think I'm done experimenting. Thanks again for putting together d12e37f.

The use-case (something like [this](https://www.techfak.uni-bielefeld.de/~jmoringe/progress.ogv)) I experimented with has the following requirements
* Establish a condition handler in all threads (transitively) created by some initial thread
* Each time the handler handles a condition, the parent thread of the current thread has to be known
* The properties have to be inherited by all ancestor threads - not just direct children of the initial thread
* All of this should work without disturbing control flow and without cooperation from library and client code (besides setting up wrappers in the initial thread)

I tried implementing this with d12e37f as well the patch attached to this pull request.

## Results for d12e37f

### Changes

I first made the following changes to make the code work (Sorry if the code was intended to work differently, I just couldn't see how):

```diff
modified   src/bordeaux-threads.lisp
@@ -106,7 +106,7 @@ FUNCTION."
     (lambda ()
       (progv (mapcar #'car specials)
           (loop for (nil . form) in specials collect (eval form))
-        (funcall (dynamic-context-wrapper) function)))))
+        (funcall (dynamic-context-wrapper context) function)))))
 
 (defstruct dynamic-context
   "The dynamic context of a thread:
@@ -130,7 +130,9 @@ FUNCTION."
                                       (dynamic-context-special-bindings
                                        *dynamic-context*))
             ,@(if wrapper `(:wrapper
-                            (funcall ,wrapper ',(dynamic-context-wrapper *dynamic-context*)))))))
+                            (let ((old (dynamic-context-wrapper *dynamic-context*))
+                                  (new   ,wrapper))
+                             (lambda (function) (funcall new (lambda () (funcall old function))))))))))
      ,@body))
 
modified   src/default-implementations.lisp
@@ -54,7 +54,7 @@ It is safe to call repeatedly."
     implementation-defined. Portable code should not depend on
     particular behaviour in this case, nor should it assign to such
     variables without first rebinding them in the new thread."
-  (%make-thread (binding-default-specials function dynamic-context)
+  (%make-thread (binding-defaults function dynamic-context)
                 (or name "Anonymous thread")))
 
 (defdfun %make-thread (function name)
```

### Client Code

The best client code I could come up with looks like this:

```cl
(defun inherit-context (set-context)
  (let ((context))
    (funcall set-context (lambda (new-value) (setf context new-value)))
    (lambda (next)
      (let ((bt:*dynamic-context* context))
        (funcall next)))))

(defun note-parent-thread (next)
  (let ((*parent-thread* (bt:current-thread)))
    (funcall next)))

(defun handle-function (function)
  (lambda (next)
    (handler-bind ((more-conditions:progress-condition
                    (lambda (condition)
                      (funcall function condition (bt:current-thread) *parent-thread*))))
      (funcall next))))

(let ((*handle-condition-function* (lambda (condition thread parent-thread)
                                     (format t "~A ~A ~A~%" thread parent-thread condition)))
      (set-context))
  (bt:with-dynamic-context (:wrapper (inherit-context (lambda (context) (setf set-context context))))
    (bt:with-dynamic-context (:wrapper #'note-parent-thread)
      (bt:with-dynamic-context (:wrapper (handle-function *handle-condition-function*))
        (funcall set-context bt:*dynamic-context*)
        (bt:make-thread
         (named-lambda :thread-1 ()
           (sleep 1)
           (more-conditions:progress :foo .1)
           (sleep 1)
           (bt:make-thread
            (named-lambda :thread-2 ()
              (more-conditions:progress :foo .3)
              (sleep 1)
              (bt:make-thread
               (lambda ()
                 (more-conditions:progress :foo .5)
                 (sleep 1)
                 (more-conditions:progress :foo .6)))
              (sleep 3)
              (more-conditions:progress :foo .4)))
           (sleep 6)
           (more-conditions:progress :foo .2))))))
  :back)
```

### Observations

* Setting up inheritance of the current dynamic context is awkward (see `set-context` hack in the above listing) because d12e37f has no  way of executing code in the parent thread before the child thread is spawned
* I couldn't make `note-parent-thread` work for the same reason
* I didn't use the `:initial-bindings` keyword parameter of `with-dynamic-context` and instead established bindings in wrapper functions because it just seemed easier. (The initial bindings semantics are a bit unclear to me, because `eval` is used and also because of the `;; Forms are evaluated in the new thread or in the calling thread?` comment)

## Results for original patch attached to this PR

### Client Code

```cl
(declaim (special *parent-thread*))

(defun inherit-wrappers (next)
  (let ((wrappers bt:*default-wrappers*))
    (lambda ()
      (let ((bt:*default-wrappers* wrappers))
        (funcall next)))))

(defun note-parent-thread (next)
  (let ((parent-thread (bt:current-thread)))
    (lambda ()
      (let ((*parent-thread* parent-thread))
        (funcall next)))))

(defun handle-function (next)
  (let ((function *handle-condition-function*))
    (lambda ()
      (handler-bind ((more-conditions:progress-condition
                      (lambda (condition)
                        (funcall function condition (bt:current-thread) *parent-thread*))))
        (let ((*handle-condition-function* function))
         (funcall next))))))

(let ((*handle-condition-function* (lambda (condition thread parent-thread)
                                     (format t "~A ~A ~A~%" thread parent-thread condition )))
      (bt:*default-wrappers*       (list* #'inherit-wrappers
                                          #'note-parent-thread
                                          #'handle-function
                                          bt:*default-wrappers*)))
  (bt:make-thread
   (named-lambda :thread-1 ()
     (sleep 1)
     (more-conditions:progress :foo .1)
     (sleep 1)
     (bt:make-thread
      (named-lambda :thread-2 ()
        (more-conditions:progress :foo .3)
        (sleep 1)
        (bt:make-thread
         (named-lambda :thread-3 ()
           (more-conditions:progress :foo .5)
           (sleep 1)
           (more-conditions:progress :foo .6)))
        (sleep 3)
        (more-conditions:progress :foo .4)))
     (sleep 6)
     (more-conditions:progress :foo .2)))
  :back)
```

### Observations

* Arguably harder to understand since all three aspects are implemented as closure-returning functions
* But achieves what is needed for the use-case

## Conclusion

I'm not sure, honestly.

I still think that "two-stage" wrappers that evaluate some code in the parent thread and some code in the child thread could be the necessary and sufficient building block for things related to dynamic context. On the other hand, I understand your concern about complexity and debuggability. Maybe one could argue that this is acceptable as such a feature will probably be used rarely and only in advanced use-cases (like initial bindings before it)?

Somewhat orthogonally to the question which variant of wrappers should be used, I suggest implementing initial bindings in terms of wrappers. This reduces the complexity and makes initial bindings completely legacy, opening the door to its removal (not without deprecation and a transition period, of course).

Sorry for drawing this out like this, I'm trying really hard to find the best solution.

sionescu commented:
I pushed all the fixes to branch context(b0253508f19a).
As for the requirements, I'm not at all convinced of them, especially of the last, that "all of this should work without disturbing control flow and without cooperation from library and client code".
And about the "two-stage" wrappers, I don't see how they're necessary or something you can't do already now.

scymtym commented:
> I pushed all the fixes to branch context(b025350).

Thanks.

> As for the requirements, I'm not at all convinced of them, especially of the last, that "all of this should work without disturbing control flow and without cooperation from library and client code".

Can you elaborate? I view this in a similar manner as special variables, non-local exist, conditions and restarts all of which work when surrounded by/wrapped around not explicitly cooperating code.

> And about the "two-stage" wrappers, I don't see how they're necessary or something you can't do already now.

For example, I cannot see a way to do the following: set up a dynamic context in the initial thread that binds the _respective_ parent thread in all ancestor threads, i.e. in
```cl
(bt:with-dynamic-context (<something that binds *parent-thread*>) 
  (bt:make-thread (named-lambda :thread-1 ()
                    (bt:make-thread (named-lambda :thread-2 ()
                                      (bt:make-thread (named-lambda :thread-3 () ...)))))))
```
the following bindings would be established:

Thread function | Value of `*parent-thread*`
----------|------------
`:thread-1` | `<initial thread>`
`:thread-2` | `:thread-1`
`:thread-3` | `:thread-2`

I'm basically ready to give up since we seem to be unable to convince each other. As a last resort, should we ask some people in #lisp for opinions?
